File: hphp/hack/src/hhbc/options_ffi.rs
// Copyright (c) Facebook, Inc. and its affiliates.
//
// This source code is licensed under the MIT license found in the
// LICENSE file in the "hack" directory of this source tree.

use hhbc_by_ref_options::Options;
use ocamlrep_ocamlpool::ocaml_ffi;

ocaml_ffi! {
    fn configs_to_json_ffi(
        jsons: Vec<compile_ffi::OcamlStr>,
        cli_args: Vec<compile_ffi::OcamlStr>,
    ) -> String {
        let jsons: Vec<_> = jsons.into_iter().rev().collect();
        Options::from_configs(&jsons, &cli_args)
            .expect("bug in deserializing Hhbc_options from Rust")
            .to_string()
    }
}

File: hphp/hack/src/hhbc/options_ffi.ml
(*
 * Copyright (c) Facebook, Inc. and its affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the "hack" directory of this source tree.
 *
 *)

external configs_to_json_ffi : string list -> string list -> string
  = "configs_to_json_ffi"

let from_configs ~(jsons : string list) ~(args : string list) : Hhbc_options.t =
  let merged = configs_to_json_ffi jsons args in
  Hhbc_options.extract_config_options_from_json
    ~init:Hhbc_options.default
    (Some (Hh_json.json_of_string merged))

let get_default () = from_configs [] []
